adafruit_logging.py

```diff
@@ -148,12 +148,84 @@ def _logRecordFactory(name, level, msg, args):
     return LogRecord(name, level, _level_for(level), msg, time.monotonic(), args)
 
 
+class Formatter:
+    """
+    Responsible for converting a LogRecord to an output string to be
+    interpreted by a human or external system.
+
+    Only implements a sub-set of CPython logging.Formatter behavior,
+    but retains all the same arguments in order to match the API.
+
+    The only init arguments currently supported are: fmt, defaults and
+    style. All others are currently ignored
+
+    The only two styles currently supported are '%' and '{'. The default
+    style is '{'
+    """
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        fmt: Optional[str] = None,
+        datefmt: Optional[str] = None,
+        style: str = "%",
+        validate: bool = True,
+        defaults: Dict = None,
+    ):
+        self.fmt = fmt
+        self.datefmt = datefmt
+        self.style = style
+        if self.style not in ("{", "%"):
+            raise ValueError(
+                "Only '%' and '{' formatting style are supported at this time."
+            )
+
+        self.validate = validate
+        self.defaults = defaults
+
+    def format(self, record: LogRecord) -> str:
+        """
+        Format the given LogRecord into an output string
+        """
+        if self.fmt is None:
+            return record.msg
+
+        vals = {
+            "name": record.name,
+            "levelno": record.levelno,
+            "levelname": record.levelname,
+            "message": record.msg,
+            "created": record.created,
+            "args": record.args,
+        }
+        if "{asctime}" in self.fmt or "%(asctime)s" in self.fmt:
+            now = time.localtime()
+            # pylint: disable=line-too-long
+            vals[
+                "asctime"
+            ] = f"{now.tm_year}-{now.tm_mon:02d}-{now.tm_mday:02d} {now.tm_hour:02d}:{now.tm_min:02d}:{now.tm_sec:02d}"
+
+        if self.defaults:
+            for key, val in self.defaults.items():
+                vals[key] = val
+
```

Comment:
Should this not set val if key is already in vals?

Reply:
Yes it should. Thank you.  I confirmed by trying `defaults={"ip": "192.168.1.188", "message": "CLOBBER"},` in with cpython `logging`. It does not end up overwriting the real log message. 

The latest commit changes the circuitpython implementation behavior to match.